### PR TITLE
added new function - modulo calculator

### PR DESCRIPTION
**Pull request by tomijava**

### CODE REVIEW

File: calculator/interfaces/Main.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.InputMismatchException;
+
+public class Main {
+
+    public static void main(String[] args) {
+        Scanner scan = new Scanner(System.in);
+        String calculatorModel = "0";
+
+        closeProgram: 
+            while(true) {
```
Comment (tomaszte) [inline calculator/interfaces/Main.java]: no need for that infinite loop, you are calling break closeProgram;  on exit anyway. Removing that while loop and calling just break would result in the same behavior.

File: calculator/interfaces/Main.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
+package calculator.interfaces;
+import java.util.Scanner;
+import java.util.InputMismatchException;
+
+public class Main {
+
+    public static void main(String[] args) {
+        Scanner scan = new Scanner(System.in);
+        String calculatorModel = "0";
```
Comment (tomaszte) [inline calculator/interfaces/Main.java]: create final variables for "0" and "1"

File: calculator/interfaces/Main.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                                case '/':
+                                    calculatorBasic.divide(Double.parseDouble(input.substring(1)));
+                                    break;
+                                case '*':
+                                    calculatorBasic.multiply(Double.parseDouble(input.substring(1)));
+                                    break;
+                                default:
+                                    System.out.println("Invalid command");
+                            }
+                        } catch (Exception exceptions) {
```
Comment (tomaszte) [inline calculator/interfaces/Main.java]: its exception not exceptions, you cant store multiple exceptions inside single Exception object.

**Review by tomaszte — CHANGES_REQUESTED** [review]

File: calculator/basic/CalculatorBasic.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+        if (this.result == (int)this.result) {
+            return Integer.toString((int)this.result);
+        }
+        else {
+            return Double.toString(this.result);
+        }
+    } 
+
+    public void add(double number) {
```
Comment (tomaszte) [inline calculator/basic/CalculatorBasic.java]: make private

File: calculator/interfaces/Calculator.java [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package calculator.interfaces;
+
+public interface Calculator {
+
+	void calculate(String number);
```
Comment (tomaszte) [inline calculator/interfaces/Calculator.java]: add toString() to interface

File: calculator/basic/CalculatorBasic.java (hunk 4 above)
Comment (tomaszte) [reply calculator/basic/CalculatorBasic.java]: wait, it is basic calculator. Isn't it merged already? Why is it shown here?

File: calculator/interfaces/BinaryCalculator.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import java.util.Scanner;
+import java.lang.Math;
+public class BinaryCalculator {
```
Comment (tomaszte) [inline calculator/interfaces/BinaryCalculator.java]: remove that file

File: calculator/basic/CalculatorBasic.java (hunk 4 above)
Comment (tomaszte) [reply calculator/basic/CalculatorBasic.java]: I see you added interface to basic package, do not edit files in basic package. That exercise is done :)

**Review by tomaszte — CHANGES_REQUESTED** [review]
Needs work.

File: calculator/interfaces/Calculator.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package calculator.interfaces;
+
+public interface Calculator {
+
+    String toString();
+    void calculate(String number);
+
```
Comment (tomaszte) [inline calculator/interfaces/Calculator.java]: remove redundant spaces

File: calculator/interfaces/CalculatorBinary.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    public CalculatorBinary() {
+        this.result = 0;
+    }
+
+    public String toString() {
+        if (result == 0) {
+            return "0";
+        }
+        else {
+            return changeToBinary();
```
Comment (tomaszte) [inline calculator/interfaces/CalculatorBinary.java]: changeToBinary() suggest that you are changing object state and it is not true.
What happens is you are getting object result as binary string so
maybe getAsBinaryString() or getResultAsBinaryString()

File: calculator/interfaces/CalculatorBinary.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+            divide(input.substring(1));
+        }
+        else if (input.charAt(0) == '*') {
+            multiply(input.substring(1));
+        }
+    }
+    private String changeToBinary() {
+        StringBuilder number = new StringBuilder();
+        int resultToConvert = this.result;
+        if (resultToConvert > 0) {
```
Comment (tomaszte) [inline calculator/interfaces/CalculatorBinary.java]: redundant if

File: calculator/interfaces/CalculatorBinary.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+                else {
+                    number.append(0);
+                }
+
+                resultToConvert /= 2;
+            }
+            number.append("-");
+            return number.reverse().toString();
+        }
+        else {
```
Comment (tomaszte) [inline calculator/interfaces/CalculatorBinary.java]: case when result is 0 is covered in toString(),remove that case from here or from ToString

File: calculator/interfaces/CalculatorBinary.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+            subtract(input.substring(1));
+        }
+        else if (input.charAt(0) == '/') {
+            divide(input.substring(1));
+        }
+        else if (input.charAt(0) == '*') {
+            multiply(input.substring(1));
+        }
+    }
+    private String changeToBinary() {
```
Comment (tomaszte) [inline calculator/interfaces/CalculatorBinary.java]: there is repeated code in that method (logic inside while's is almost the same) You could get absolute copy of result (absolute number means - sign is removed -5 is 5 -3 is 3, 3 is still 3 etc.) and on that absolute value do that while logic (while (resultToConvert > 0)) and after that you can add minus sign if needed and call reverse().toString()

File: calculator/interfaces/CalculatorBinary.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+                resultToConvert /= 2;
+            }
+            number.append("-");
+            return number.reverse().toString();
+        }
+        else {
+            return "0";
+        }
+    } 
+    private void add (String input) {
```
Comment (tomaszte) [inline calculator/interfaces/CalculatorBinary.java]: interesting solution! but you already have changeToBinary, you can reuse that make changeToBinary take input parameter so you could write this.result += changeToBinary(number)

File: calculator/interfaces/CalculatorBinary.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    }
+    private void divide (String input) {
+        StringBuilder number = new StringBuilder(input);
+        int divider = 0;
+        for (int index = 0; index < number.length(); index++) {
+            if (number.charAt(index) == '1') {
+                divider += Math.pow(2, ((number.length() - 1) - index));
+            }
+        }
+        this.result /= divider;
```
Comment (tomaszte) [inline calculator/interfaces/CalculatorBinary.java]: what happens if divider is 0?

File: calculator/interfaces/CalculatorDateAndTime.java [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+		this.zone = ZonedDateTime.now(ZoneId.of(zone.idZone));
+	}
+	public String toString () {
+		return zone.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
+	}
+    public String whenSunday () {
+        return "Sunday is " + zone.with(TemporalAdjusters.next(DayOfWeek.SUNDAY)).format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+    }
+    public void showSecondsSince (String input) {
+        Pattern yearMonthDay = Pattern.compile("^showSecondsSince(\\d{1,4})\\s?([1-9]|1[0-2])?\\s?([1-9]|1[0-9]|2[0-9]|3[01])?$");
```
Comment (tomaszte) [inline calculator/interfaces/CalculatorDateAndTime.java]: please describe what is that regex accepting and add few examples of correct input. (In code as comment, do that for every long more complicated regex)

File: calculator/interfaces/CalculatorDateAndTime.java [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+        }
+        else {
+            System.out.println("Bad date");
+        }
+    }
+    public void printHelp() {
+        System.out.println("All commands \nwhenSunday = show when will be the nearest Sunday \nshowSecondsSince [enter date example: 1991 12 12] = show how much seconds elapsed from date \nchangeCity = changes city(timezone)");
+        System.out.println("Add/delete = example + 5y 3m - add 5 years and 3 months \ny = years, m = months, d = day, h = hour, min = minute, s = second");
+    }
+	public void calculate (String number) {
```
Comment (tomaszte) [inline calculator/interfaces/CalculatorDateAndTime.java]: tabs to spaces

File: calculator/interfaces/CalculatorDateAndTime.java (hunk 14 above)
Comment (tomaszte) [reply calculator/interfaces/CalculatorDateAndTime.java]: and compile pattern once, it is rather, compiling pattern is rather slow, im not sure though :)

Comment (tomaszte) [reply calculator/interfaces/CalculatorDateAndTime.java]: rename pattern to better describe what it is, maybe showSecondsSincePattern ?

File: calculator/interfaces/CalculatorDateAndTime.java [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.TemporalAdjusters;
+import java.time.DayOfWeek;
+import java.time.Duration;
+import java.time.Year;
+import java.time.YearMonth;
+import java.time.LocalDate;
+public class CalculatorDateAndTime implements Calculator {
+
+	private final Pattern pattern = Pattern.compile("^([\\+\\-])\\s*((\\d*)y)?\\s*((\\d*)m)?\\s*((\\d*)d)?\\s*((\\d*)h)?\\s*((\\d*)min)?\\s*((\\d*)s)?$");
```
Comment (tomaszte) [inline calculator/interfaces/CalculatorDateAndTime.java]: change variable name to describe what is inside

File: calculator/interfaces/CalculatorDateAndTime.java [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+        }
+    }
+    public void printHelp() {
+        System.out.println("All commands \nwhenSunday = show when will be the nearest Sunday \nshowSecondsSince [enter date example: 1991 12 12] = show how much seconds elapsed from date \nchangeCity = changes city(timezone)");
+        System.out.println("Add/delete = example + 5y 3m - add 5 years and 3 months \ny = years, m = months, d = day, h = hour, min = minute, s = second");
+    }
+	public void calculate (String number) { 
+		Matcher matcher = pattern.matcher(number);
+		if (matcher.find()) {
+			if (matcher.group(1).equals("+")) {
```
Comment (tomaszte) [inline calculator/interfaces/CalculatorDateAndTime.java]: a lot of repeated code, maybe first get sign to variable; + or -, next get date into variable (years, days, months, minutes together) and then do something like zone.plusDate(date)/zone.minusDate(date)

Comment (tomaszte) [reply calculator/interfaces/CalculatorDateAndTime.java]: if(matcher.group(2) != null && matcher.group(3) != null)
put magic numbers like this to final static variables i.e.
static final PATTERN_GROUP_YEARS_ID = 2;
static final PATTERN_GROUP_YEARS_VALUE_ID = 3;
...
if(matcher.group(PATTERN_GROUP_YEARS_ID) != null && matcher.group(PATTERN_GROUP_YEARS_VALUE_ID) != null)

File: calculator/interfaces/CalculatorDateAndTime.java (hunk 14 above)
Comment (tomaszte) [reply calculator/interfaces/CalculatorDateAndTime.java]: put every pattern string to static final variable
static final SHOW_SECONDS_SINCE_PATTERN_STR = "^showSecondsSince(\\d{1,4})\\s?([1-9]|1[0-2])?\\s?([1-9]|1[0-9]|2[0-9]|3[01])?$";
that pattern is also used in main method, reuse that variable there

File: calculator/interfaces/CalculatorDateAndTime.java [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+        else if (matcher.group(1) != null) {
+            Year year = Year.of(Integer.valueOf(matcher.group(1)).intValue());
+            secondsSince = Duration.between(zone, zone.minusYears(zone.getYear() - year.getValue()));
+            System.out.println("Seconds: " + secondsSince.getSeconds());
+        }
+        else {
+            System.out.println("Bad date");
+        }
+    }
+    public void printHelp() {
```
Comment (tomaszte) [inline calculator/interfaces/CalculatorDateAndTime.java]: decide if you put space between function name and parameters
printHelp() or printHelp ()
calculate (String number) or calculate(String number)?
make every method follow same format

File: calculator/interfaces/Main.java [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+            else {
+                System.out.println("Bad command");
+            }
+        }
+        while (true) {
+            if (calculatorDateAndTime != null) {
+                System.out.println(calculatorDateAndTime.toString());
+                String input = StringHelper.removeWhitespace(scan.nextLine());
+                Matcher matcher = pattern.matcher(input);
+                Pattern patternForShowSince = Pattern.compile("^showSecondsSince(\\d{1,4})\\s?([1-9]|1[0-2])?\\s?([1-9]|1[0-9]|2[0-9]|3[01])?$");
```
Comment (tomaszte) [inline calculator/interfaces/Main.java]: compile once, not in loop

File: calculator/interfaces/Main.java [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+                pattern = Pattern.compile("[\\+\\-\\/\\*]{1}[0-1]*");
+                break;
+            }
+            else if (calculatorModel.equals(dateAndTimeCalculator)) {
+                System.out.println("Choose your time zone");
+
+                for (Zone id : Zone.values()) {
+                    System.out.println(id.number + ": " + id);
+                }
+                int city = scan.nextInt();
```
Comment (tomaszte) [inline calculator/interfaces/Main.java]: that is cityId not city itself, object City containing city information could be named city.

File: calculator/interfaces/Main.java [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+                }
+                else if (input.equals("changeCity")) {
+                    for (Zone id : Zone.values()) {
+                        System.out.println(id.number + " " + id);
+                    }
+                    System.out.println("Enter the city number");
+                    while (true) {
+                        int city = scan.nextInt();
+                        scan.nextLine();
+                        if (city >= 1 && city <= 8) {
```
Comment (tomaszte) [inline calculator/interfaces/Main.java]: move that logic as Zone enum method like checkIfValidZoneId or something like that.

File: calculator/interfaces/Main.java [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+        System.out.println("Enter 2 for modulo calculator");
+        System.out.println("Enter 3 for binary calculator");
+        System.out.println("Enter 4 for Date and Time calculator");
+
+        while(true) {
+            calculatorModel = scan.nextLine();
+
+            if (calculatorModel.equals(basicCalculator)) {
+                calculator = new CalculatorBasic();
+                pattern = Pattern.compile("[\\+\\-\\*\\/]{1}[0-9]+");
```
Comment (tomaszte) [inline calculator/interfaces/Main.java]: add that as class field, like inputPattern/inputMatcher and initialize in constructor

File: calculator/interfaces/Main.java [hunk 23]
```diff
@@ -0,0 +1,141 @@
+package calculator.interfaces;
+import java.util.Scanner;
+import java.util.InputMismatchException;
+import java.util.regex.*;
+import java.time.ZoneId;
+
+public class Main {
+
+    public static final String basicCalculator = "1";
```
Comment (tomaszte) [inline calculator/interfaces/Main.java]: public static final variables are names in convention like BASIC_CALCULATOR_ID

Comment (tomaszte) [reply calculator/interfaces/Main.java]: fix them all plx

File: calculator/interfaces/Main.java [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+                        else {
+                            System.out.println("Bad command enter correct number");
+                        }
+                    }
+                }
+                else if (input.equals("whenSunday")) {
+                    System.out.println(calculatorDateAndTime.whenSunday());
+
+                }
+                else if (matcherForShowSince.matches()) {
```
Comment (tomaszte) [inline calculator/interfaces/Main.java]: not bad name, but type (Matcher) is added at end of the variable so showSecondsSinceMatcher

File: calculator/interfaces/Main.java [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+                }
+                else {
+                    System.out.println("Bad command");
+                    calculatorDateAndTime.printHelp();
+                }               
+            }
+            else {
+                System.out.println(calculator.toString());
+                String input = StringHelper.removeWhitespace(scan.nextLine());
+                Matcher matcher = pattern.matcher(input);
```
Comment (tomaszte) [inline calculator/interfaces/Main.java]: describe what matcher is that

**Review by tomaszte — CHANGES_REQUESTED** [review]

File: calculator/interfaces/Calculator.java (hunk 7 above)
Comment (tomijava) [reply calculator/interfaces/Calculator.java]: fixed

File: calculator/interfaces/CalculatorBinary.java (hunk 8 above)
Comment (tomijava) [reply calculator/interfaces/CalculatorBinary.java]: fixed

File: calculator/interfaces/CalculatorBinary.java (hunk 10 above)
Comment (tomijava) [reply calculator/interfaces/CalculatorBinary.java]: fixed

File: calculator/interfaces/Main.java (hunk 23 above)
Comment (tomijava) [reply calculator/interfaces/Main.java]: fixed

**Pull request closed without merging**